In [99]:
# Data manipulation and formatting configuration
import pandas as pd
pd.set_option('display.max_columns', None)

# Numerical operations
import numpy as np

# Advanced machine learning imputation setup
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
imputer_iter = IterativeImputer(max_iter= 10, random_state= 42)

# Notebook environment cleanup
import warnings
warnings.filterwarnings('ignore')

## 01. Exploración de datos.

In [ ]:
# Load the customer flight activity dataset and inspect its initial shape (rows and columns)
df_flight_activity = pd.read_csv("../data/Customer Flight Activity.csv")
df_flight_activity.shape

(405624, 10)

In [ ]:
# Load the customer loyalty history dataset and inspect its initial shape
df_loyalty = pd.read_csv("../data/Customer Loyalty History.csv")
df_loyalty.shape

(16737, 16)

### DF1: Flight Activity

In [100]:
df_flight_activity.sample(5)

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
377021,377742,2018,11,0,0,0,0,0.00,0,0
320972,817070,2017,2,0,0,0,0,0.00,0,0
39567,409768,2017,3,0,0,0,0,0.00,0,0
285944,184456,2018,2,6,4,10,1370,171.25,355,29
239783,129429,2018,2,9,3,12,2268,244.08,0,0


In [6]:
df_flight_activity.info()

<class 'pandas.DataFrame'>
RangeIndex: 405624 entries, 0 to 405623
Data columns (total 10 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Year                         405624 non-null  int64  
 2   Month                        405624 non-null  int64  
 3   Flights Booked               405624 non-null  int64  
 4   Flights with Companions      405624 non-null  int64  
 5   Total Flights                405624 non-null  int64  
 6   Distance                     405624 non-null  int64  
 7   Points Accumulated           405624 non-null  float64
 8   Points Redeemed              405624 non-null  int64  
 9   Dollar Cost Points Redeemed  405624 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 30.9 MB


In [ ]:
# No negative numbers detected
(df_flight_activity < 0).any().any()

np.False_

In [8]:
# Year Verification: the unique values correspond to calendar years
df_flight_activity['Year'].unique()

array([2017, 2018])

In [9]:
# Month verification: numbers must be between 1 and 12
df_flight_activity['Month'].between(1, 12, inclusive='both').all()

np.True_

*DataFrame Info Observations:*

- No negative values detected in the DF.
- No null values detected in the DF.
- Column names follow a consistent style: they include spaces and use Title Case.
- "Year" and "Month" are stored as int instead of datetime. Numbers correspond to year and month values.
    * Decision: Maintained as int since the dataset lacks full date strings (missing days). Keeping them as integers ensures lighter performance and smoother DataFrame operations.
    * ⬜ Check columns with high zero counts to verify if they contain other values.
- Significant difference in total entries between DataFrames: he same customer can appear multiple times across different flights.
    * ⬜ Count the number of unique customers in the dataset.
    * ⬜ Check for overall duplicate records in the DataFrame.
- Reviewed the study requirements and analyzed the available columns in the DataFrame: 
    * ⬜ Assess whether "Points Redeemed" and "Dollar Cost" are useful features.

In [10]:
df_flight_activity['Loyalty Number'].unique().shape

(16737,)

In [11]:
df_flight_activity.duplicated().sum()

np.int64(1864)

In [12]:
df_flight_activity['Flights with Companions'].unique().shape

(12,)

In [13]:
df_flight_activity['Points Redeemed'].unique().shape

(587,)

In [14]:
df_flight_activity['Dollar Cost Points Redeemed'].unique().shape

(49,)

- ✅ Count the number of unique customers in the dataset:
    * Update: Validated; the unique customer count matches the rows in secondary DataFrame.
- ✅ Check for overall duplicate records in the DataFrame:
    * Update: Duplicates verified. 
- ✅ Check columns with high zero counts to verify if they contain other values:
    * Update: Verified; all columns contain unique values other than zero.
- ✅ Assess whether 'Points Redeemed' and 'Dollar Cost' are useful features.
    * Update: Evaluated both columns; although they contain valid non-null entries, they are non-informative for the scope of this analysis. Therefore, a decision was made to drop both features.

- Final Data Cleaning Checklist:
    * ⬜ Remove the duplicate records from the DataFrame.
    * ⬜ Drop 'Points Redeemed' and 'Dollar Cost' columns from the DataFrame.

### DF2: Loyalty History

In [16]:
df_loyalty.head()

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
1,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN
2,429460,Canada,British Columbia,Vancouver,V6E 3D9,Male,College,NaN,Single,Star,3839.75,Standard,2014,7,2018.0,1.0
3,608370,Canada,Ontario,Toronto,P1W 1K4,Male,College,NaN,Single,Star,3839.75,Standard,2013,2,NaN,NaN
4,530508,Canada,Quebec,Hull,J8Y 3Z5,Male,Bachelor,103495.0,Married,Star,3842.79,Standard,2014,10,NaN,NaN


In [89]:
df_loyalty.info()

<class 'pandas.DataFrame'>
RangeIndex: 16737 entries, 0 to 16736
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Loyalty Number      16737 non-null  int64  
 1   Country             16737 non-null  str    
 2   Province            16737 non-null  str    
 3   City                16737 non-null  str    
 4   Postal Code         16737 non-null  str    
 5   Gender              16737 non-null  str    
 6   Education           16737 non-null  str    
 7   Salary              12499 non-null  float64
 8   Marital Status      16737 non-null  str    
 9   Loyalty Card        16737 non-null  str    
 10  CLV                 16737 non-null  float64
 11  Enrollment Type     16737 non-null  str    
 12  Enrollment Year     16737 non-null  int64  
 13  Enrollment Month    16737 non-null  int64  
 14  Cancellation Year   2067 non-null   float64
 15  Cancellation Month  2067 non-null   float64
dtypes: float64(4), 

In [ ]:
# Check for negative values
loyalty_numerical_cols = df_loyalty.select_dtypes(include='number')
(loyalty_numerical_cols < 0).any().any()

np.True_

In [90]:
# Check for null values
loyalty_null_pct = round(df_loyalty.isnull().sum()/df_loyalty.shape[0]*100, 2)
loyalty_null_pct = loyalty_null_pct[loyalty_null_pct>0]
loyalty_null_pct

Salary                25.32
Cancellation Year     87.65
Cancellation Month    87.65
dtype: float64

*DataFrame Info Observations:*

- Negative values detected:
    * ⬜ Check the numerical columns to verify where are the negative values.
- Column names follow a consistent style: they include spaces and use Title Case. Also consistent with the other DataFrame.
    * ⬜ Verify the number of unique customers in the dataset.
- The "Salary" column presents a significant missing data rate of 25%.
    * ⬜ Add to the null value handling pipeline.
- "Cancellation Year" and "Cancellation Month" exhibit a high missingness rate of 87.65%. This indicates they represent active customers, meaning null imputation is not required. Additionally, the columns are stored as float instead of int, because of the presence of null values.
    * ⬜ Assess whether these columns are useful features for the scope of the study.

In [20]:
# No negative numbers in Loyalty Number
df_loyalty['Loyalty Number'].min()

np.int64(100018)

In [21]:
# Negative numbers detected in Salary
df_loyalty['Salary'].min()

np.float64(-58486.0)

In [ ]:
# Negative numbers are not repeated
negative_values = df_loyalty[df_loyalty['Salary']< 0]['Salary']
print(len(negative_values))
print(negative_values.value_counts().sort_index(ascending=True))

20
Salary
-58486.0    1
-57297.0    1
-49830.0    1
-49001.0    1
-47310.0    1
-46683.0    1
-46470.0    1
-46303.0    1
-45962.0    1
-43234.0    1
-39503.0    1
-34079.0    1
-31911.0    1
-26322.0    1
-19332.0    1
-19325.0    1
-17534.0    1
-12497.0    1
-10605.0    1
-9081.0     1
Name: count, dtype: int64


In [ ]:
# Check lower bounds of positive salaries to verify continuity with negative values
df_loyalty['Salary'].sort_values()[20:]

3243     15609.0
15197    16628.0
14896    16743.0
3471     16853.0
7152     17166.0
          ...   
16721        NaN
16727        NaN
16731        NaN
16732        NaN
16734        NaN
Name: Salary, Length: 16717, dtype: float64

In [22]:
# No negative numbers in CLV
df_loyalty['CLV'].min()

np.float64(1898.01)

In [23]:
# Enrollment Year verification: No negatives. The unique values correspond to calendar years.
df_loyalty['Enrollment Year'].unique()

array([2016, 2014, 2013, 2012, 2015, 2018, 2017])

In [ ]:
# Enrollment Month verification: No negatives. Numbers must be between 1 and 12
df_loyalty['Enrollment Month'].between(1, 12, inclusive='both').all()

np.True_

In [25]:
# Cancellation Year verification: No negatives. The unique values correspond to calendar years.
df_loyalty['Cancellation Year'].dropna().unique()

array([2018., 2015., 2017., 2014., 2016., 2013.])

In [ ]:
# Cancellation Month verification: No negatives. Numbers must be between 1 and 12
df_loyalty['Cancellation Month'].dropna().between(1, 12, inclusive='both').all()

np.True_

In [27]:
# Verification that the customer count matches the flight DataFrame
df_loyalty['Loyalty Number'].value_counts().sum()

np.int64(16737)

In [ ]:
#No duplicates founded in the second DataFrame
df_loyalty.duplicated().sum()

np.int64(0)

* ✅ Check the numerical columns to verify where are the negative values:
    - Update: negative values detected in Salary. 
* ✅ Verify the number of unique customers in the dataset:
    - Update: Confirmed; zero duplicate entries found.
* ✅ Assess whether these columns are useful features for the scope of the study:
    - Update: For the purpose of this study, the 'Cancellation Year' and 'Cancellation Month' columns are not required. However, a new feature will be created to preserve the customer's active status.

- Final Data Cleaning Checklist:
    * ⬜ Clean the negative numbers from 'Salary' column.
    * ⬜ Create new columns with customer's status.
    * ⬜ Drop the columns 'Cancellation Year' and 'Cancellation Month' from the DataFrame.
    * ⬜ Handle null values in the 'Salary' column.

### Merging the DataFrames

In [ ]:
df_merge = df_flight_activity.merge(df_loyalty, on= ['Loyalty Number'], how = 'left')

In [38]:
df_merge.shape

(403760, 22)

In [64]:
df_merge.head()

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Customer Active
0,480934,2017,1,0,0,0,0,0.0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,True
1,480934,2017,2,3,0,3,2823,282.0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,True
2,480934,2017,3,0,0,0,0,0.0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,True
3,480934,2017,4,0,0,0,0,0.0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,True
4,480934,2018,10,6,2,8,3352,335.0,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,True


### Data Cleaning

**Data Cleaning Checklist:**

* ⬜ Create new columns with customer's status.
* ⬜ Drop 'Points Redeemed', 'Dollar Cost', 'Cancellation Year' and 'Cancellation Month' columns from the DataFrame.
* ⬜ Remove the duplicate records from the DataFrame.
* ⬜ Clean the negative numbers from 'Salary' column.
* ⬜ Handle null values in the 'Salary' column.

In [32]:
# Create a new column indicating whether a customer is active or not by handling the null values.
df_merge['Customer Active'] = df_merge['Cancellation Year'].isnull()

In [33]:
# Dropping columns that will not be used in the study.
df_merge.drop(columns=['Dollar Cost Points Redeemed','Points Redeemed','Cancellation Year','Cancellation Month'], inplace=True)

In [ ]:
df_merge.duplicated().sum()

np.int64(1864)

In [ ]:
# Dropping duplicated rows
df_merge.drop_duplicates(inplace=True)

In [37]:
df_merge.duplicated().sum()

np.int64(0)

In [54]:
# The presence of negative values in the 'Salary' column is nonsensical. 
# Since negative numbers are not repeated, it is assumed to be a typo; therefore, it was decided to convert these values to positive numbers.

df_merge['Salary'] = df_merge['Salary'].abs()

In [68]:
df_merge['Salary'].min()

np.float64(9081.0)

In [ ]:
# The missing value percentage remains identical after merging. 
# Since it exceeds 20%, advanced imputation techniques will be applied.
round(df_merge['Salary'].isnull().sum() / len(df_merge) * 100, 2)

np.float64(25.33)

In [94]:
# Select all numerical columns to serve as predictors for the model
merge_numerical_cols = df_merge.select_dtypes(include=['number']).columns

# Fit and transform the numerical columns (returns a NumPy array)
imputed_values = imputer_iter.fit_transform(df_merge[merge_numerical_cols])

# Convert the resulting array back into a temporary DataFrame
df_temp = pd.DataFrame(imputed_values, columns=merge_numerical_cols, index=df_merge.index)

# Safely overwrite ONLY the 'Salary' column in our main DataFrame
df_merge['Salary'] = df_temp['Salary']

print(f"Remaining missing values in 'Salary': {df_merge['Salary'].isnull().sum()}")

Remaining missing values in 'Salary': 0


In [ ]:
# 
df_merge.info()

<class 'pandas.DataFrame'>
Index: 403760 entries, 0 to 405623
Data columns (total 22 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Loyalty Number           403760 non-null  int64  
 1   Year                     403760 non-null  int64  
 2   Month                    403760 non-null  int64  
 3   Flights Booked           403760 non-null  int64  
 4   Flights with Companions  403760 non-null  int64  
 5   Total Flights            403760 non-null  int64  
 6   Distance                 403760 non-null  int64  
 7   Points Accumulated       403760 non-null  float64
 8   Country                  403760 non-null  str    
 9   Province                 403760 non-null  str    
 10  City                     403760 non-null  str    
 11  Postal Code              403760 non-null  str    
 12  Gender                   403760 non-null  str    
 13  Education                403760 non-null  str    
 14  Salary              

In [98]:
df_merge.sample(5)

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Customer Active
80350,738890,2018,11,0,0,0,0,0.00,Canada,Newfoundland,St. John's,A1C 6H9,Male,College,79095.433353,Single,Aurora,11836.30,Standard,2014,12,False
229382,732808,2018,3,3,0,3,2619,326.25,Canada,Quebec,Montreal,H2T 2J6,Male,College,78779.154640,Single,Nova,18708.76,Standard,2016,10,True
96751,348994,2017,7,4,0,4,3864,386.00,Canada,Ontario,Thunder Bay,K8T 5M5,Male,Doctor,165418.000000,Married,Aurora,25390.55,Standard,2016,3,True
80098,336281,2017,11,9,6,15,3675,367.00,Canada,Alberta,Banff,T4V 1D4,Female,Master,118863.000000,Divorced,Aurora,11790.50,Standard,2017,5,True
88655,594200,2018,12,0,0,0,0,0.00,Canada,Ontario,Ottawa,K1F 2R2,Female,College,78435.362400,Married,Aurora,15563.37,Standard,2018,12,True


**Data Cleaning Checklist (Updated):**

* ✅ Create new columns with customer's status.
    - Update: Successfully engineered and extracted status-related features.
* ✅ Drop 'Points Redeemed', 'Dollar Cost', 'Cancellation Year', and 'Cancellation Month' columns from the DataFrame.
    - Update: Removed unnecessary and redundant features to optimize the dataset.
* ✅ Remove the duplicate records from the DataFrame.
    - Update: Dropped duplicate rows to ensure data integrity before merging.
* ✅ Clean the negative numbers from 'Salary' column.
    - Update: Identified sign errors (typos) and corrected them by converting negative values into positive absolute numbers.
* ✅ Handle null values in the 'Salary' column.
    - Update: Addressed missing data (exceeding 20%) by implementing an advanced machine learning imputation technique (IterativeImputer/MICE with random_state=42), preserving data distribution and variance.

In [101]:
# Save the final cleaned and merged DataFrame to a new CSV file
df_merge.to_csv("../data/Customer_Flight_Loyalty_Clean.csv", index=False)